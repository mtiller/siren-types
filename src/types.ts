/**
 * These types are based on the JSON Schema found at:
 *
 * https://gist.github.com/kevinswiber/14477759858a768d2809326ca4300d26
 */

// This is a alias for a Siren entity where the type of the properties has a default
export type Siren<P extends {} = any> = Entity<P>;

// Shorthand for the mouthful that is EmbeddedRepresentationSubEntity.
export type EmbeddedEntity<P> = EmbeddedRepresentationSubEntity<P>;

// This is another short-hand
export type EmbeddedLink = EmbeddedLinkSubEntity;

export type Properties = {} | undefined;

/**
 * This is the definition of a Siren entity.  Note that everything is optional
 */
export interface Entity<P extends Properties = {}> {
    // class: Describes the nature of an entity's content based on the current representation.
    // Possible values are implementation-dependent and should be documented.
    class?: string[];
    // title - Descriptive text about the entity.
    title?: string;
    // A set of key-value pairs that describe the state of an entity.
    properties: P;
    // entities - A collection of related sub-entities. If a sub-entity contains an href value,
    // it should be treated as an embedded link. Clients may choose to optimistically load embedded
    // links. If no href value exists, the sub-entity is an embedded entity representation that
    // contains all the characteristics of a typical entity. One difference is that a sub-entity
    // MUST contain a rel attribute to describe its relationship to the parent entity.
    entities?: SubEntity<any>[];
    // actions - A collection of actions; actions show available behaviors an entity exposes.
    actions?: Action[];
    // links - A collection of items that describe navigational links, distinct from entity
    // relationships. Link items should contain a `rel` attribute to describe the relationship
    // and an `href` attribute to point to the target URI. Entities should include a link `rel` to `self`.
    links?: Link[];
}

/**
 * Entities in Siren come in two flavors.  Personally, I've never used the
 * embedded link sub-entity, but I include it here for completeness.
 */
export type SubEntity<P extends {} = {}> = EmbeddedLinkSubEntity | EmbeddedRepresentationSubEntity<P>;

/**
 * The aforementioned (but never actually used by me) EmbeddedLinkSubEntity.
 */
export interface EmbeddedLinkSubEntity {
    // required - rel and href
    // class - Describes the nature of an entity's content based on the current
    // representation. Possible values are implementation-dependent and should be documented.
    class?: string[];
    // rel - Defines the relationship of the sub-entity to its parent, per Web Linking (RFC5899).
    // required and cannot be empty.
    rel: string[];
    // href - The URI of the linked sub-entity.
    href: string;
    // type - Defines media type of the linked sub-entity, per Web Linking (RFC5899).
    type?: string;
    // title - Descriptive text about the entity.
    title?: string;
}

/**
 * An embedded sub-entity.  It is just like any other Siren representation
 * except that it MUST include a relation as well.
 */
export interface EmbeddedRepresentationSubEntity<P> extends Entity<P> {
    // rel - Defines the relationship of the sub-entity to its parent, per Web Linking (RFC5899).
    // required and cannot be empty.
    rel: string[];
}

/**
 * Actions show available behaviors an entity exposes.
 */
export interface Action {
    // name - A string that identifies the action to be performed.
    // Action names MUST be unique within the set of actions for an entity.
    // The behaviour of clients when parsing a Siren document that violates
    // this constraint is undefined.
    name: string;
    // class - Describes the nature of an action based on
    // the current representation. Possible values are
    // implementation-dependent and should be documented.
    class?: string[];
    // method - An enumerated attribute mapping to a protocol method. For HTTP,
    // these values may be GET, PUT, POST, DELETE, or PATCH. As new methods are
    // introduced, this list can be extended. If this attribute is omitted, GET
    // should be assumed.  (default: "GET")
    method?: "GET" | "PUT" | "POST" | "DELETE" | "PATCH" | "HEAD" | "OPTIONS";
    // href - The URI of the action.
    href: string;
    // title - Descriptive text about the action.
    title?: string;
    // type - The encoding type for the request. When omitted and the fields
    // attribute exists, the default value is `application/x-www-form-urlencoded`.
    type?: string;
    // fields - A collection of fields.
    fields?: Field[];
}

export interface Field {
    // name - A name describing the control. Field names MUST be unique within the
    // set of fields for an action. The behaviour of clients when parsing a Siren
    // document that violates this constraint is undefined.
    name: string;
    // type - The input type of the field. This is a subset of the input types specified by HTML5.
    type?:
        | "hidden"
        | "text"
        | "search"
        | "tel"
        | "url"
        | "email"
        | "password"
        | "datetime"
        | "date"
        | "month"
        | "week"
        | "time"
        | "datetime-local"
        | "number"
        | "range"
        | "color"
        | "checkbox"
        | "radio"
        | "file";
    // title - Textual annotation of a field. Clients may use this as a label.
    title?: string;
    // value - A value assigned to the field.  May be a scalar value or a list of value objects.
    value?: string | number | FieldValueObject[];
}

/**
 * Value objects represent multiple selectable field values. Use
 * in conjunction with field `"type" = "radio"` and `"type" = "checkbox"`
 * to express that zero, one or many out of several possible values may be sent back to the server.
 */
export interface FieldValueObject {
    // value - Possible value for the field.
    value: string | number;
    // title - Textual description of a field value.
    title?: string;
    // selected - A value object with a `"selected" = true` attribute
    // indicates that this value should be considered preselected by
    // the client. When missing, the default value is `false`.
    selected?: boolean;
}

/**
 * Links represent navigational transitions.
 */
export interface Link {
    // class - Describes aspects of the link based on the current representation.
    // Possible values are implementation-dependent and should be documented.
    class?: string[];
    // title - Text describing the nature of a link.
    title?: string;
    // rel - Defines the relationship of the link to its entity, per Web Linking (RFC5988).
    rel: string[];
    // href - The URI of the linked resource.
    href: string;
    // type - Defines media type of the linked resource, per Web Linking (RFC5988).
    type?: string;
}
