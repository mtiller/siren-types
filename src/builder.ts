import { Siren, Field } from "./types";

interface Builder<T> {
  asJSON(): Siren<T>;
}

export interface PartialLink {
  class?: string[];
  title?: string;
  type?: string;
}

export interface PartialAction {
  class?: string[];
  method?: "GET" | "PUT" | "POST" | "DELETE" | "PATCH" | "HEAD" | "OPTIONS";
  title?: string;
  type?: string;
  fields?: Field[];
}

export class SirenResponse<T> implements Builder<T> {
  private resp: Siren<T>;
  constructor(def: T, prev?: Siren<T>) {
    if (def) {
      this.resp = {
        ...(prev || {}),
        properties: def
      } as Siren<T>;
    } else {
      this.resp = { ...(prev || {}) } as Siren<T>;
    }
  }
  addProperty<K extends keyof T>(key: K, val: T[K]): SirenResponse<T> {
    let props: T = this.resp.properties;
    if (props) {
      (props as any)[key] = val;
    }
    return new SirenResponse<T>(props, this.resp);
  }
  properties(props: T): SirenResponse<T> {
    return new SirenResponse<T>(props, this.resp);
  }
  link(rel: string[], href: string, rest?: PartialLink): SirenResponse<T> {
    const resp: Siren<T> = {
      ...this.resp,
      links: [
        ...(this.resp.links || []),
        {
          ...(rest || {}),
          rel: rel,
          href: href
        }
      ]
    };
    return new SirenResponse<T>(this.resp.properties, resp);
  }
  action(name: string, href: string, rest?: PartialAction): SirenResponse<T> {
    const resp: Siren<T> = {
      ...this.resp,
      actions: [
        ...(this.resp.actions || []),
        {
          ...(rest || {}),
          name: name,
          href: href
        }
      ]
    };
    return new SirenResponse<T>(this.resp.properties, resp);
  }
  asJSON(): Siren<T> {
    return this.resp;
  }
}
