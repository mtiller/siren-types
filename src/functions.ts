import {
  Entity,
  SubEntity,
  Properties,
  EmbeddedLinkSubEntity,
  EmbeddedRepresentationSubEntity,
  EmbeddedRepresentationSubEntityWithoutProperties,
  EmbeddedRepresentationSubEntityWithProperties
} from "./types";

/** Shorthand function */
export function isA(
  e: Entity<any>,
  className: string,
  ...additionalClasses: string[]
): boolean {
  const names = [className, ...additionalClasses];
  return names.some(name => e.class.indexOf(name) >= 0);
}

/**
 * Fetch (all) self relations.  Only the href is included here because these
 * might come from a Link, EmbeddedLink or EmbeddedEntity (very unlikely).
 * @param s
 */
export function collectSelves(e: Entity<any>): string[] {
  let ret: string[] = [];
  if (e.links) {
    for (let i = 0; i < e.links.length; i++) {
      // istanbul ignore else
      if (e.links[i].rel.indexOf("self") >= 0) {
        ret.push(e.links[i].href);
      }
    }
  }
  if (e.entities) {
    for (let i = 0; i < e.entities.length; i++) {
      const entity = e.entities[i];
      if (isEmbeddedLink(entity)) {
        // istanbul ignore else
        if (entity.rel.indexOf("self") >= 0) {
          ret.push(entity.href);
        }
      } else {
        // From the spec:
        //
        // Root entities and sub-entities that are embedded representations
        // SHOULD contain a links collection with at least one item contain
        // a rel value of self and an href attribute with a value of the
        // entity's URI.
        // istanbul ignore else
        if (entity.rel.indexOf("self") >= 0) {
          ret = [...ret, ...collectSelves(entity)];
        }
      }
    }
  }
  return ret;
}

/**
 * If you use this function in a conditional statement, the TypeScript compiler
 * will know to narrow the type of the argument based on the result.
 */
export function isEmbeddedLink<P extends Properties | undefined>(
  s:
    | EmbeddedLinkSubEntity
    | EmbeddedRepresentationSubEntityWithoutProperties
    | EmbeddedRepresentationSubEntityWithProperties<P>
): s is EmbeddedLinkSubEntity {
  return s.hasOwnProperty("href");
}

/**
 * If you use this function in a conditional statement, the TypeScript compiler
 * will know to narrow the type of the argument based on the result.
 */
export function isEmbeddedRepr<P>(
  s: SubEntity<P>
): s is EmbeddedRepresentationSubEntity<P> {
  return !s.hasOwnProperty("href");
}
