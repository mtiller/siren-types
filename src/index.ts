export * from "./constants";
export * from "./types";
export * from "./functions";
export * from "./normalized";
export * from "./builder";
