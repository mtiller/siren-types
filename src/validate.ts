import {
  EmbeddedLinkSubEntity,
  Properties,
  SubEntity,
  EmbeddedRepresentationSubEntity,
  Siren
} from "./types";

import { createIs, createAssertType } from "typescript-is";

/**
 * If you use this function in a conditional statement, the TypeScript compiler
 * will know to narrow the type of the argument based on the result.
 */
export function isEmbeddedLink<P extends Properties | undefined>(
  s: SubEntity<P>
): s is EmbeddedLinkSubEntity {
  return s.hasOwnProperty("href");
}

/**
 * If you use this function in a conditional statement, the TypeScript compiler
 * will know to narrow the type of the argument based on the result.
 */
export function isEmbeddedRepr<P>(
  s: SubEntity<P>
): s is EmbeddedRepresentationSubEntity<P> {
  return !s.hasOwnProperty("href");
}

export const isSiren = createIs<Siren>();
export const validateSiren = createAssertType<Siren>();
