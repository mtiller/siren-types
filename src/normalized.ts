import {
  Entity,
  Link,
  Action,
  EmbeddedLinkSubEntity,
  Properties,
  EmbeddedRepresentationSubEntity,
  EmbeddedRepresentationSubEntityWithProperties,
  EmbeddedRepresentationSubEntityWithoutProperties,
  SubEntity,
} from "./types";
import { isEmbeddedLink } from "./functions";

/**
 * A slight variation on the Siren type where everything is now required.  Title
 * is a special case because I didn't want to use `""` as the default value so I
 * allow null in that case.  For everything else, empty objects and arrays suffice.
 */
export interface NormalizedEntity<
  P extends Properties | undefined = undefined
> {
  class: string[];
  title: string | null;
  properties: P;
  entities: NormalizedSubEntity[];
  actions: NormalizedAction[];
  links: Link[];
}

/**
 * This function normalizes an existing Siren Entity.
 *
 * Note this doesn't normalize properties because we cannot know how to
 * normalize them.  So we just hope they are what we expect.
 */
export function normalize<P extends Properties | undefined = undefined>(
  e: Entity<P>,
  defaultProps: P
): NormalizedEntity<P> {
  let props: P = e.properties === undefined ? undefined : defaultProps;
  if (props) {
    return {
      class: e.class || [],
      title: e.title || null,
      properties: props,
      entities: e.entities
        ? e.entities.map((x) => normalizeEmbedded(x as SubEntity))
        : [],
      actions: e.actions ? e.actions.map((x) => normalizeAction(x)) : [],
      links: e.links || [],
    };
  } else {
    return {
      class: e.class || [],
      title: e.title || null,
      entities: e.entities
        ? e.entities.map((x) => normalizeEmbedded(x as SubEntity))
        : [],
      actions: e.actions ? e.actions.map((x) => normalizeAction(x)) : [],
      links: e.links || [],
    } as NormalizedEntity<P>;
  }
}

export function normalizeEmbeddedLinkSubEntity(
  e: EmbeddedLinkSubEntity
): NormalizedEmbeddedLinkSubEntity {
  return {
    class: e.class || [],
    rel: e.rel,
    href: e.href,
    type: e.type || "application/octet-stream",
    title: e.title || null,
  };
}

export function normalizeEmbeddedRepresentationSubEntity<
  P extends Properties | undefined
>(
  e: EmbeddedRepresentationSubEntity<P>
): NormalizedEmbeddedRepresentationSubEntity<P> {
  return {
    class: e.class || [],
    rel: e.rel,
    title: e.title || null,
    properties: e.properties as P,
    entities: e.entities
      ? e.entities.map((x) => normalizeEmbedded(x as SubEntity))
      : [],
    actions: e.actions ? e.actions.map((x) => normalizeAction(x)) : [],
    links: e.links || [],
  };
}

export function normalizeEmbedded<P extends Properties | undefined>(
  e: SubEntity<P>
): NormalizedSubEntity<P> {
  if (isEmbeddedLink(e)) {
    return normalizeEmbeddedLinkSubEntity(e);
  }
  return normalizeEmbeddedRepresentationSubEntity(e);
}

/**
 * Normalized forms of the subentity types
 */
export type NormalizedSubEntity<P extends Properties | undefined = undefined> =
  | NormalizedEmbeddedLinkSubEntity
  | NormalizedEmbeddedRepresentationSubEntity<P>;

export interface NormalizedEmbeddedLinkSubEntity {
  class: string[];
  rel: string[];
  href: string;
  type: string;
  title: string | null;
}

/**
 * Normalized version of an embedded representation
 */
export interface NormalizedEmbeddedRepresentationSubEntity<P>
  extends NormalizedEntity<P> {
  rel: string[];
}

export type NormalizedAction = Required<Action>;

export function normalizeAction(a: Action): NormalizedAction {
  return {
    name: a.name,
    class: a.class || [],
    method: a.method || "GET",
    href: a.href,
    title: a.title || a.name,
    type: a.type || "application/x-www-form-urlencoded",
    fields: a.fields || [],
  };
}
