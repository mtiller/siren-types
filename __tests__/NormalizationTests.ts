import { Siren, normalize, Action, normalizeAction, SubEntity } from "../src";

describe("Test normalizing functions", () => {
  it("should normalize an empty Entity", () => {
    const base: Siren = {};
    const normal = normalize(base, undefined);
    expect(normal).toEqual({
      class: [],
      title: null,
      actions: [],
      links: [],
      entities: []
    });
  });

  it("should normalize an Entity with a class", () => {
    const base: Siren = {
      class: ["empty"]
    };
    const normal = normalize(base, undefined);
    expect(normal).toEqual({
      class: ["empty"],
      title: null,
      actions: [],
      links: [],
      entities: []
    });
  });

  it("should normalize an Action", () => {
    const base: Action = {
      name: "create",
      href: "/#"
    };
    const normal = normalizeAction(base);
    expect(normal).toEqual({
      name: base.name,
      class: [],
      method: "GET",
      href: base.href,
      title: base.name,
      type: "application/x-www-form-urlencoded",
      fields: []
    });
  });

  it("should normalize subentities", () => {
    const base: Siren = {
      entities: [
        {
          class: ["items", "collection"],
          rel: ["http://x.io/rels/order-items"],
          href: "http://api.x.io/orders/42/items"
        },
        {
          class: ["info", "customer"],
          rel: ["http://x.io/rels/customer"],
          properties: {
            customerId: "pj123",
            name: "Peter Joseph"
          },
          links: [{ rel: ["self"], href: "http://api.x.io/customers/pj123" }]
        },
        {
          rel: ["http://x.io/rels/customer"],
          properties: {
            customerId: "pj124",
            name: "Peter Joseph"
          },
          links: [{ rel: ["self"], href: "http://api.x.io/customers/pj124" }]
        },
        {
          class: ["items", "collection"],
          rel: ["http://x.io/rels/order-items"],
          entities: [
            {
              rel: ["http://x.io/rels/order-items"],
              href: "http://api.x.io/orders/42/items"
            },
            {
              rel: ["http://x.io/rels/order-items"],
              href: "http://api.x.io/orders/y/items"
            }
          ],
          actions: [
            {
              name: "create",
              title: "Create Item",
              href: "http://api.x.io/orders/42/items"
            }
          ]
        } as SubEntity<undefined>
      ],
      actions: [
        {
          name: "create",
          title: "Create Item",
          href: "http://api.x.io/orders/42/items"
        }
      ]
    };
    const normal = normalize(base, undefined);
    expect(normal).toMatchSnapshot();
  });
});
