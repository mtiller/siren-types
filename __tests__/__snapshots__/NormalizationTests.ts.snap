// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test normalizing functions should normalize subentities 1`] = `
Object {
  "actions": Array [
    Object {
      "class": Array [],
      "fields": Array [],
      "href": "http://api.x.io/orders/42/items",
      "method": "GET",
      "name": "create",
      "title": "Create Item",
      "type": "application/x-www-form-urlencoded",
    },
  ],
  "class": Array [],
  "entities": Array [
    Object {
      "class": Array [
        "items",
        "collection",
      ],
      "href": "http://api.x.io/orders/42/items",
      "rel": Array [
        "http://x.io/rels/order-items",
      ],
      "title": null,
      "type": "application/octet-stream",
    },
    Object {
      "actions": Array [],
      "class": Array [
        "info",
        "customer",
      ],
      "entities": Array [],
      "links": Array [
        Object {
          "href": "http://api.x.io/customers/pj123",
          "rel": Array [
            "self",
          ],
        },
      ],
      "properties": Object {
        "customerId": "pj123",
        "name": "Peter Joseph",
      },
      "rel": Array [
        "http://x.io/rels/customer",
      ],
      "title": null,
    },
    Object {
      "actions": Array [],
      "class": Array [],
      "entities": Array [],
      "links": Array [
        Object {
          "href": "http://api.x.io/customers/pj124",
          "rel": Array [
            "self",
          ],
        },
      ],
      "properties": Object {
        "customerId": "pj124",
        "name": "Peter Joseph",
      },
      "rel": Array [
        "http://x.io/rels/customer",
      ],
      "title": null,
    },
    Object {
      "actions": Array [
        Object {
          "class": Array [],
          "fields": Array [],
          "href": "http://api.x.io/orders/42/items",
          "method": "GET",
          "name": "create",
          "title": "Create Item",
          "type": "application/x-www-form-urlencoded",
        },
      ],
      "class": Array [
        "items",
        "collection",
      ],
      "entities": Array [
        Object {
          "class": Array [],
          "href": "http://api.x.io/orders/42/items",
          "rel": Array [
            "http://x.io/rels/order-items",
          ],
          "title": null,
          "type": "application/octet-stream",
        },
        Object {
          "class": Array [],
          "href": "http://api.x.io/orders/y/items",
          "rel": Array [
            "http://x.io/rels/order-items",
          ],
          "title": null,
          "type": "application/octet-stream",
        },
      ],
      "links": Array [],
      "properties": undefined,
      "rel": Array [
        "http://x.io/rels/order-items",
      ],
      "title": null,
    },
  ],
  "links": Array [],
  "title": null,
}
`;
